### PR TITLE
[Merged by Bors] - chore: change the definition of `List.finRange`

### PR DESCRIPTION
**Pull request by kim-em**
François Dorais has been [working](https://github.com/leanprover-community/batteries/pull/1055) on upstreaming `List.finRange`, but wants to change the definition at the same time. That was running into difficulties, which hopefully are resolved here.

### CODE REVIEW

File: Mathlib/Data/List/Defs.lean [hunk 1]
```diff
@@ -488,9 +488,12 @@ theorem length_mapAccumr₂ :
 
 end MapAccumr
 
+/- #adaptation_note: this attribute should be removed after Mathlib moves to v4.15.0-rc1. -/
+set_option allowUnsafeReducibility true in
+attribute [semireducible] Fin.foldr.loop
```
Comment (eric-wieser) [inline Mathlib/Data/List/Defs.lean:493]: Does this apply globally, or just for the current file?

File: Mathlib/Data/List/Range.lean [hunk 2]
```diff
@@ -42,33 +43,43 @@ theorem chain_range_succ (r : ℕ → ℕ → Prop) (n a : ℕ) :
   rw [range_succ_eq_map, chain_cons, and_congr_right_iff, ← chain'_range_succ, range_succ_eq_map]
   exact fun _ => Iff.rfl
 
+-- Verify that `finRange` is semireducible.
+example : finRange 3 = [0, 1, 2] := rfl
```
Comment (eric-wieser) [inline Mathlib/Data/List/Range.lean:47]: Can this go directly after `def finRange`, or are the numerals not available there?

File: Mathlib/Data/List/Range.lean [hunk 3]
```diff
@@ -42,33 +43,43 @@ theorem chain_range_succ (r : ℕ → ℕ → Prop) (n a : ℕ) :
...
+
+theorem finRange_eq_pmap_range (n : ℕ) : finRange n = (range n).pmap Fin.mk (by simp) := by
+  apply List.ext_getElem <;> simp [finRange]
+
 @[simp]
-theorem finRange_zero : finRange 0 = [] :=
-  rfl
```
Comment (eric-wieser) [inline Mathlib/Data/List/Range.lean:47]: This change stops this working as a `dsimp` lemma; does the `rfl` proof still work?

File: Mathlib/Data/List/Defs.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Data/List/Defs.lean:493]: I also worry that this PR might have accumulated a bunch of proof uglification before we discovered this fix; is that the case, or do I have the chronology wrong?

File: Mathlib/Data/List/Range.lean (hunk 3 above)
Comment (kim-em) [reply Mathlib/Data/List/Range.lean:47]: Restored.

File: Mathlib/Data/List/Range.lean (hunk 2 above)
Comment (kim-em) [reply Mathlib/Data/List/Range.lean]: Done.

File: Mathlib/Data/List/Defs.lean (hunk 1 above)
Comment (kim-em) [reply Mathlib/Data/List/Defs.lean:493]: The uglification seems to still be necessary at the moment, and I'm not really sure why it is happening. It is only at the level of having to redo some simp only blocks, however.

File: Mathlib/NumberTheory/Modular.lean [hunk 4]
```diff
@@ -207,17 +207,20 @@ theorem tendsto_lcRow0 {cd : Fin 2 → ℤ} (hcd : IsCoprime (cd 0) (cd 1)) :
...
+      of_apply, cons_val', cons_val_zero, empty_val', cons_val_fin_one, lcRow0Extend_apply,
       LinearMap.GeneralLinearGroup.coeFn_generalLinearEquiv, GeneralLinearGroup.coe_toLinear,
-      val_planeConformalMatrix, neg_neg, mulVecLin_apply, cons_val_one, head_cons, of_apply,
-      Fin.mk_zero, Fin.mk_one]
-  · convert congr_arg (fun n : ℤ => (-n : ℝ)) g.det_coe.symm using 1
-    simp only [f₁, mulVec, dotProduct, Fin.sum_univ_two, Matrix.det_fin_two, Function.comp_apply,
-      Subtype.coe_mk, lcRow0Extend_apply, cons_val_zero,
+      val_planeConformalMatrix, neg_neg, mulVecLin_apply, mulVec, dotProduct, Fin.sum_univ_two,
+      cons_val_one, head_cons, mB, f₁]
+  · -- FIXME: I have horribly broken this proof, and need to using `g.det_coe` a second time. :-(
```
Comment (jcommelin) [inline Mathlib/NumberTheory/Modular.lean:216]: Is this `FIXME` still accurate?

File: Mathlib/NumberTheory/Modular.lean [hunk 5]
```diff
@@ -207,17 +207,20 @@ theorem tendsto_lcRow0 {cd : Fin 2 → ℤ} (hcd : IsCoprime (cd 0) (cd 1)) :
...
+    convert congr_arg (fun n : ℤ => (-n : ℝ)) g.det_coe.symm using 1
```
Comment (kim-em) [inline Mathlib/NumberTheory/Modular.lean:217]:
```suggestion
  · convert congr_arg (fun n : ℤ => (-n : ℝ)) g.det_coe.symm using 1
```

**Review by jcommelin — COMMENTED** [review]
Thanks :tada: 

If CI passes, please remove the label `awaiting-CI` and merge this yourself, by adding a comment `bors r+`.

bors d+

File: Mathlib/Data/List/Defs.lean (hunk 1 above)
Comment (kim-em) [reply Mathlib/Data/List/Defs.lean:493]: > Does this apply globally, or just for the current file?

This is global.

**Comment by kim-em** [comment]
bors merge

**Pull request closed without merging**